Meta/Lagom/Tools/CodeGenerators/LibPublicSuffix/GeneratePublicSuffixData.cpp

```diff
@@ -0,0 +1,186 @@
+    args_parser.add_option(public_suffix_list_path, "Path to the public suffix list", "public-suffix-list-path", 'p', "public-suffix-list-path");
+    args_parser.parse(arguments);
+
+    auto identifier_data = TRY(open_file(public_suffix_list_path, Core::File::OpenMode::Read));
+
+    auto generated_header_file = TRY(Core::File::open(generated_header_path, Core::File::OpenMode::Write));
+    auto generated_implementation_file = TRY(Core::File::open(generated_implementation_path, Core::File::OpenMode::Write));
+
+    TRY(generate_header_file(*identifier_data, *generated_header_file));
+    TRY(generate_implementation_file(*identifier_data, *generated_implementation_file));
+
+    return 0;
+}
+
+ErrorOr<void> generate_header_file(Core::InputBufferedFile&, Core::File& file)
+{
+    StringBuilder builder;
+    SourceGenerator generator { builder };
+    generator.append(R"~~~(
+#pragma once
+
+#include <AK/DeprecatedString.h>
+#include <AK/Forward.h>
+#include <AK/Trie.h>
+
+namespace PublicSuffix {
+
+class PublicSuffixData {
+protected:
+    PublicSuffixData();
+
+public:
+    PublicSuffixData(PublicSuffixData const&) = delete;
+    PublicSuffixData& operator=(PublicSuffixData const&) = delete;
+
+    static PublicSuffixData* the()
+    {
+        static PublicSuffixData* s_the;
+        if (!s_the)
+            s_the = new PublicSuffixData;
+        return s_the;
+    }
+
+    ErrorOr<Optional<String>> get_public_suffix(StringView string);
+
+private:
+    Trie<char, DeprecatedString> m_dictionary;
+};
+
+} // namespace PublicSuffix
+
+)~~~");
+
+    TRY(file.write_until_depleted(generator.as_string_view().bytes()));
+    return {};
+}
+
+ErrorOr<void> generate_implementation_file(Core::InputBufferedFile& input, Core::File& file)
+{
+    StringBuilder builder;
+    SourceGenerator generator { builder };
+    generator.append(R"~~~(
+#include <LibPublicSuffix/PublicSuffixData.h>
+#include <AK/Vector.h>
+#include <AK/String.h>
+
+namespace PublicSuffix {
+
+static Vector<StringView> s_public_suffixes {)~~~");
+
+    Array<u8, 1024> buffer {};
+
+    while (TRY(input.can_read_line())) {
+        auto line = TRY(input.read_line(buffer));
+
+        if (line.starts_with("//"sv) || line.is_empty())
+            continue;
+
+        generator.set("line", line);
+        generator.append(R"~~~(
+    {".@line@"sv},)~~~");
+    }
+
+    generator.append(R"~~~(
+};
+
+PublicSuffixData::PublicSuffixData()
+    : m_dictionary('/', "")
+{
+    // FIXME: Reduce the depth of this trie
+    for (auto str : s_public_suffixes) {
+        auto view = str.split_view("."sv);
+        view.reverse();
+
+        StringBuilder builder;
+        builder.join("."sv, view);
+        auto val = builder.string_view();
+
+        auto it = val.begin();
+        MUST(m_dictionary.insert(it, val.end(), val, [](auto& parent, auto& it) -> Optional<DeprecatedString> { 
+            return DeprecatedString::formatted("{}{}", parent.metadata_value(), *it);
+        }));
+    }
```

Comment:
Might be an option to do this rearrangement loop in the generator, so that it doesn't have to happen at runtime?